Service/Product/FacetBuilder.php

```diff
@@ -0,0 +1,242 @@
+    public const FACET_ATTRIBUTES_CATEGORY_ID = 'categoryIds';
+
+    public const FACET_KEY_ATTRIBUTE_NAME = 'attribute';
+    public const FACET_KEY_SEARCHABLE = 'searchable';
+
+    public const FACET_SEARCHABLE_SEARCHABLE = '1';
+    public const FACET_SEARCHABLE_NOT_SEARCHABLE = '2';
+    public const FACET_SEARCHABLE_FILTER_ONLY = '3';
+
+    // Local raw facet cache, indexed by $storeId
+    protected array $facets = [];
+
+    public function __construct(
+        protected ConfigHelper                            $configHelper,
+        protected StoreManagerInterface                   $storeManager,
+        protected CurrencyHelper                          $currencyManager,
+        protected GroupCollection                         $groupCollection,
+        protected GroupExcludedWebsiteRepositoryInterface $groupExcludedWebsiteRepository,
+    )
+    {}
+
+    /**
+     * @param int $storeId
+     * @return string[]
+     * @throws LocalizedException
+     * @throws NoSuchEntityException
+     */
+    public function getAttributesForFaceting(int $storeId): array
+    {
+        return array_map(
+            function($facet) {
+                return $this->decorateAttributeForFaceting($facet);
+            },
+            $this->getRawFacets($storeId)
+        );
+    }
+
+    /**
+     * @param int $storeId
+     * @return array<string, array>|null
+     * @throws LocalizedException
+     * @throws NoSuchEntityException
+     */
+    public function getRenderingContent(int $storeId): ?array
+    {
+        $facets = $this->getRawFacets($storeId);
+        if (empty($facets)) {
+            return null;
+        }
+        $attributes = $this->extractFacetAttributeNames($facets);
+        return [
+            'facetOrdering' => [
+                'facets' => [
+                    'order' => $attributes
+                ],
+                'values' => $this->getRenderingContentValues($attributes)
+            ],
+        ];
+    }
+
+    /**
+     * @param array<array<string, mixed>> $facets
+     * @return string[]
+     */
+    protected function extractFacetAttributeNames(array $facets): array
+    {
+        return array_map(
+            function($facet) {
+                return $facet[self::FACET_KEY_ATTRIBUTE_NAME];
+            },
+            $facets
+        );
+    }
+
+    /**
+     * @param string[] $facets
+     * @return array<string, array>
+     */
+    protected function getRenderingContentValues(array $facets): array
+    {
+        return array_combine(
+            array_map(
+                function(string $attribute) {
+                    if ($attribute === self::FACET_ATTRIBUTE_CATEGORIES) {
+                        $attribute = self::FACET_ATTRIBUTE_CATEGORIES . '.level0';
+                    }
+                    return $attribute;
+                },
+                $facets
+            ),
+            array_fill(0, count($facets), [ 'sortRemainingBy' => 'alpha' ])
```

Comment:
Do we plan to make this configurable in the admin in the future or only with backend events ?

Reply:
Good question! We certainly could make it configurable. But yes as of right now it would require using our backend events (or a preference class if they really want to customize things). 

I'll make a separate task as we'd need to not only alter the backend but add to the admin grid UI for configuring the facets. It also doesn't really apply to everything (like pricing - the `sortRemainingBy` doesn't really seem to matter based on my tests). I also was trying to keep the new feature simple and less dependent on Magento. At some point I'd prefer to make Algolia the source of truth and not our local configuration for this kind of stuff but that's a big lift for a future time.